azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -59,95 +62,158 @@ def __init__(self, wire_client, full_goal_state=False, base_incarnation=None):
         """
         uri = GOAL_STATE_URI.format(wire_client.get_endpoint())
         self.xml_text = wire_client.fetch_config(uri, wire_client.get_header())
-        xml_doc = parse_doc(self.xml_text)
-
-        self.incarnation = findtext(xml_doc, "Incarnation")
-        self.expected_state = findtext(xml_doc, "ExpectedState")
-        role_instance = find(xml_doc, "RoleInstance")
+        self._xml_doc = parse_doc(self.xml_text)
+
+        self._ext_config_retriever = ext_config_retriever
+        self._wire_client = wire_client
+
+        # Cached properties
+        # Implementation note: cached_property was not used because it considers a None value to mean the property
+        # wasn't retrieved
+        self._ext_conf = None
+        self._ext_conf_retrieved = False
+        self._hosting_env = None
+        self._hosting_env_retrieved = False
+        self._shared_conf = None
+        self._shared_conf_retrieved = False
+        self.certs = None
+        self._remote_access = None
+        self._remote_access_retrieved = False
+        self._artifacts_profile_blob_url = None
+        self._status_upload_blob_url = None
+        self._status_upload_blob_type = None
+        self._ext_conf_properties_retrieved = False
+
+        self.incarnation = findtext(self._xml_doc, "Incarnation")
+        self.expected_state = findtext(self._xml_doc, "ExpectedState")
```

Comment:
is self.expected_state used at all?

Reply:
I didn't find any uses, so I removed it, though I don't think I added it.